### PR TITLE
Flax pipeline pndm

### PR DESCRIPTION
**Pull request by pcuenca**
WIP for PNDM scheduler support in Flax pipeline. Includes #577 and #581 (which I need for testing).

### CODE REVIEW

File: src/diffusers/configuration_utils.py [hunk 1]
```diff
@@ -154,15 +154,25 @@ def from_config(cls, pretrained_model_name_or_path: Union[str, os.PathLike], ret
...
+            init_dict["dtype"] = unused_kwargs.pop("dtype")
+
+        # Return model and optionally state and/or unused_kwargs
         model = cls(**init_dict)
+        return_tuple = (model,)
+
+        # Some components (Flax schedulers) have a state.
+        if getattr(cls, "has_state", False):       # Check for "create_state" in model instead?
+            state = model.create_state()
+            return_tuple += (state,)
```
Comment (patil-suraj) [inline src/diffusers/configuration_utils.py:170]: We should make this a bit more robust

1. Check if it's a scheduler i.e instance of `SchedulerMixin` if yes
2. Check if the name starts with `Flax`

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 2]
```diff
@@ -172,7 +172,7 @@ def __call__(
             self.unet.in_channels,
         )
         if latents is None:
-            latents = jax.random.normal(prng_seed, shape=latents_shape, dtype=self.dtype)
+            latents = jax.random.normal(prng_seed, shape=latents_shape, dtype=jnp.float32)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:175]: can we use the models `dtype` here ?

File: src/diffusers/schedulers/scheduling_ddim_flax.py [hunk 3]
```diff
@@ -104,6 +104,7 @@ class FlaxDDIMScheduler(SchedulerMixin, ConfigMixin):
             `set_alpha_to_one=False`, to make the last step use step 0 for the previous alpha product, as done in
             stable diffusion.
     """
+    has_state = True
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_ddim_flax.py:107]: Think this should be a  read-only `property` instead of class variable so it can't be modified by users.

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 4]
```diff
@@ -112,7 +113,8 @@ class FlaxPNDMScheduler(SchedulerMixin, ConfigMixin):
             `set_alpha_to_one=False`, to make the last step use step 0 for the previous alpha product, as done in
             stable diffusion.
     """
-
+    has_state = True
+
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:117]: same comment as above

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 5]
```diff
@@ -223,73 +233,77 @@ def step(
...
-        cur_sample = state.cur_sample if state.cur_sample is not None else sample
-
-        prev_sample = self._get_prev_sample(cur_sample, timestep, prev_timestep, model_output)
-        state = state.replace(counter=state.counter + 1)
-
-        if not return_dict:
-            return (prev_sample, state)
+        return self.step_plms(
+            state=state, model_output=model_output, timestep=timestep, sample=sample, return_dict=return_dict
+        )
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:238]: what's the issue with `prk` step ?

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 6]
```diff
@@ -330,29 +344,85 @@ def step_plms(
...
+            )            
+
+        def counter_other(state: PNDMSchedulerState):
+            ets = state.ets.at[3].set(model_output)
+            next_model_output = (1 / 24) * (55 * ets[3] - 59 * ets[2] + 37 * ets[1] - 9 * ets[0])
+
+            ets = ets.at[0].set(ets[1])
+            ets = ets.at[1].set(ets[2])
+            ets = ets.at[2].set(ets[3])
+
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:410]: Think if/else should also work here. Did not work ? what was the issue ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py (hunk 2 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:175]: `self.dtype` didn't work for me when using `bfloat16`. The latents generated in the denoising loop are always computed in `float32`, so they don't match the previous type if it was something else.

File: src/diffusers/schedulers/scheduling_pndm_flax.py (hunk 6 above)
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_pndm_flax.py:439]: Resolved offline after reviewing https://github.com/patil-suraj/stable-diffusion-jax/pull/5 and https://github.com/patil-suraj/stable-diffusion-jax/pull/8.

**Comment by pcuenca** [comment]
This is ready for review, I just synced with `main`. There are almost more commits than line changes lol, because I kept changes from other PRs that I needed to test.

I tested a generation with the current Flax pipeline using the `FlaxPNDMScheduler` and it worked fine, so we may want to set it as the default for Stable Diffusion after we verify everything's correct. In that regard, I tested the `plms` steps extensively when I was working in Suraj's repo, and I haven't tested the `prk` steps at all – just checked that they run. Perhaps we can use @kashif's tests to verify.

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 7]
```diff
@@ -196,8 +197,11 @@ def set_timesteps(self, state: PNDMSchedulerState, num_inference_steps: int) ->
 
         return state.replace(
             timesteps=jnp.concatenate([state.prk_timesteps, state.plms_timesteps]).astype(jnp.int64),
-            ets=jnp.array([]),
             counter=0,
+            # Will be zeros, not really empty
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:201]: (nit) then why not use `zeros`

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 8]
```diff
@@ -227,22 +231,32 @@ def step(
...
+        if self.config.skip_prk_steps:
+            prev_sample, state = self.step_plms(
+                state=state, model_output=model_output, timestep=timestep, sample=sample
             )
         else:
-            return self.step_plms(
-                state=state, model_output=model_output, timestep=timestep, sample=sample, return_dict=return_dict
+            prev_sample, state = jax.lax.switch(
+                jnp.where(state.counter < len(state.prk_timesteps), 0, 1),
+                [self.step_prk, self.step_plms],
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:241]: (nit)
```suggestion
                (self.step_prk, self.step_plms),
```

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 9]
```diff
@@ -266,42 +280,70 @@ def step_prk(
...
+            return state.replace(cur_model_output=state.cur_model_output + 1 / 3 * model_output), model_output
+
+        def remainder_3(state: PNDMSchedulerState, model_output: jnp.ndarray, ets_at: int):
             model_output = state.cur_model_output + 1 / 6 * model_output
-            state = state.replace(cur_model_output=0)
+            return state.replace(cur_model_output=jnp.zeros_like(state.cur_model_output)), model_output
+
+        state, model_output = jax.lax.switch(
+            state.counter % 4,
+            [remainder_0, remainder_1, remainder_2, remainder_3],
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:311]: (nit)
```suggestion
            (remainder_0, remainder_1, remainder_2, remainder_3),
```

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 10]
```diff
@@ -266,42 +280,70 @@ def step_prk(
...
+        #     state = state.replace(cur_model_output=state.cur_model_output + 1 / 3 * model_output)
+        # elif (self.counter - 2) % 4 == 0:
+        #     state = state.replace(cur_model_output=state.cur_model_output + 1 / 3 * model_output)
+        # elif (self.counter - 3) % 4 == 0:
+        #     model_output = state.cur_model_output + 1 / 6 * model_output
+        #     state = state.replace(cur_model_output=0)
 
         # cur_sample should not be `None`
-        cur_sample = state.cur_sample if state.cur_sample is not None else sample
+        # cur_sample = state.cur_sample if state.cur_sample is not None else sample
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:333]: should be deleted before merging

**Review by patil-suraj — APPROVED** [review]
Very cool @pcuenca , just left tiny nits.

And yes, we can set this as default scheduler, maybe in a follow-up PR.

Thanks a lot for working on this :)

**Comment by patrickvonplaten** [comment]
Should we merge this one?

**Comment by pcuenca** [comment]
> Should we merge this one?

I think I addressed all the comments, so it should be ok to merge.

We still need to verify the numerical outputs of the PRK steps (not used in Stable Diffusion afaik). I'll try to do it today.

**Comment by patrickvonplaten** [comment]
Merging then! Think we don't need to verify prk as it's Flax. Let's only dive into it if people ask for it
